File: web-app/src/app/service/label.service.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

import { Label } from '../pojo/Label';
import { Message } from '../pojo/Message';
import { Page } from '../pojo/Page';

const label_uri = '/label';

@Injectable({
  providedIn: 'root'
})
export class LabelService {
  constructor(private http: HttpClient) {}

  public loadLabels(
    search: string | undefined,
    type: number | undefined,
    pageIndex: number,
    pageSize: number
  ): Observable<Message<Page<Label>>> {
    pageIndex = pageIndex ? pageIndex : 0;
    pageSize = pageSize ? pageSize : 8;
    // HttpParams is unmodifiable, so we need to save the return value of append/set
    let httpParams = new HttpParams();
    httpParams = httpParams.appendAll({
      pageIndex: pageIndex,
      pageSize: pageSize
    });
    if (type != undefined) {
      httpParams = httpParams.append('type', type);
    }
    if (search != undefined && search != '' && search.trim() != '') {
      httpParams = httpParams.append('search', search.trim());
    }
    const options = { params: httpParams };
    return this.http.get<Message<Page<Label>>>(label_uri, options);
  }

  public newLabel(body: Label): Observable<Message<any>> {
    return this.http.post<Message<any>>(label_uri, body);
  }

  public editLabel(body: Label): Observable<Message<any>> {
    return this.http.put<Message<any>>(label_uri, body);
  }

  public deleteLabels(ids: Set<number>): Observable<Message<any>> {
    let httpParams = new HttpParams();
    ids.forEach(id => {
      // HttpParams is unmodifiable, so we need to save the return value of append/set
      // Method append can append same key, set will replace the previous value
      httpParams = httpParams.append('ids', id);
    });
    const options = { params: httpParams };
    return this.http.delete<Message<any>>(label_uri, options);
  }
}

File: hertzbeat-manager/src/main/java/org/apache/hertzbeat/manager/service/LabelService.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.hertzbeat.manager.service;

import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.apache.hertzbeat.common.entity.manager.Label;
import org.springframework.data.domain.Page;

/**
 * Label service
 */
public interface LabelService {

    /**
     * new Label
     * @param label label
     */
    void addLabel(Label label);

    /**
     * update label
     * @param label label
     */
    void modifyLabel(Label label);

    /**
     * get label page list
     * @param search        label content search
     * @param type          label type
     * @param pageIndex     List current page
     * @param pageSize      Number of list pagination
     * @return label
     */
    Page<Label> getLabels(String search, Byte type, int pageIndex, int pageSize);

    /**
     * delete labels
     * @param ids label id list
     */
    void deleteLabels(HashSet<Long> ids);

    /**
     * Identifies new labels from the given entry that are not yet in the database
     * @param originLabels k-v entry
     * @return new labels
     */
    List<Label> determineNewLabels(Set<Map.Entry<String, String>> originLabels);
}
